composition/internal/controller/composition_controller.go

```
@@ -0,0 +1,415 @@
+
+// /
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/reconcile
+func (r *CompositionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Got a new request!", "request", req)
+
+	var composition compositionv1alpha1.Composition
+	if err := r.Client.Get(ctx, req.NamespacedName, &composition); err != nil {
+		logger.Error(err, "unable to fetch Composition")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Grab status for comparison later
+	oldStatus := composition.Status.DeepCopy()
+
+	// Try updating status before returning
+	defer func() {
+		if !reflect.DeepEqual(oldStatus, composition.Status) {
+			newStatus := composition.Status.DeepCopy()
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				nn := types.NamespacedName{Namespace: composition.Namespace, Name: composition.Name}
+				err := r.Client.Get(ctx, nn, &composition)
+				if err != nil {
+					return err
+				}
+				composition.Status = *newStatus.DeepCopy()
+				return r.Client.Status().Update(ctx, &composition)
+			})
+			if err != nil {
+				logger.Error(err, "unable to update Composition status")
+			}
+		}
+	}()
+
+	logger = logger.WithName(composition.Name).WithName(fmt.Sprintf("%d", composition.Generation))
+
+	composition.Status.ClearCondition(compositionv1alpha1.Error)
+	logger.Info("Validating Compostion object")
+	if !composition.Validate() {
+		logger.Info("Validation Failed")
+		return ctrl.Result{}, fmt.Errorf("Validation failed")
```

Should be saying what the problem was.

In fact, probably `composition.Validate` should return an error rather than a boolean?

---

Actually see other comment about just dropping this

---

Ack. Added to TODO.md for future follow up.